src/safe_pointers.h

```diff
@@ -0,0 +1,58 @@
+/* -*- c++ -*- ----------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://www.lammps.org/, Sandia National Laboratories
+   LAMMPS development team: developers@lammps.org
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+#ifndef LMP_SAFE_POINTERS_H
+#define LMP_SAFE_POINTERS_H
+
+// collection of smart pointers for specific purposes
+
+#include <cstdio>
+
+namespace LAMMPS_NS {
+
+/** Class to automatically close a FILE pointer when a class instance goes out of scope.
+
+\verbatim embed:rst
+
+Drop in replacement for ``FILE *``. Use as ``SafeFilePtr fp;`` instead of
+``FILE *fp = nullptr;`` and there is no more need to explicitly call
+``fclose(fp)``.
+
+\endverbatim
+*/
+class SafeFilePtr {
+ public:
+  SafeFilePtr() : fp(nullptr) {};
+  SafeFilePtr(const SafeFilePtr &) = delete;
+  SafeFilePtr(const SafeFilePtr &&) = delete;
+  SafeFilePtr &operator=(const SafeFilePtr &) = delete;
+
+  ~SafeFilePtr()
+  {
+    if (fp) fclose(fp);
+  }
+
+  SafeFilePtr &operator=(FILE *_fp)
+  {
+    if (fp && (fp != _fp)) fclose(fp);
+    fp = _fp;
+    return *this;
+  }
+  operator FILE *() const { return fp; }
+
+ private:
+  FILE *fp;
+};
```

Comment:
Here is a replacement that just uses `std::unique_ptr` (needs `#include <memory>` on top). See https://godbolt.org/z/enK4jYY57 for validation code.
```suggestion
class SafeFilePtr {
 public:
  SafeFilePtr(FILE * f) : fp(f, std::fclose) {}
  SafeFilePtr() : SafeFilePtr(nullptr) {}
  SafeFilePtr(SafeFilePtr &) = delete;
  SafeFilePtr& operator=(SafeFilePtr&) = delete;
  
  operator FILE*() const {
    return fp.get();
  }

 private:
  std::unique_ptr<FILE, int (*)(FILE*)> fp;
};
```

Reply:
Thanks for the suggestion. It is quite sophisticated. In this case, however, I prefer the more explicit version for better readability and accessibility for less experienced C++ programmers.
I am open to revisit this for other use cases of the same programming pattern. After all, I've named the new header file "safe_pointers.h" because I expect that other, similar will be added over time.

Reply:
:roll_eyes: